src/decoder/lattice-incremental-decoder.cc

```diff
@@ -0,0 +1,1295 @@
+        } else { // final_arc
+          // record final_arc in this chunk for the step 3.2 in the next call
+          KALDI_ASSERT(clat.Final(arc.nextstate) != CompactLatticeWeight::Zero());
+          final_arc_list_.push_back(
+              pair<int32, size_t>(state_appended, aiter.Position()));
+        }
+      }
+    }
+  }
+  // sanity check, remove it later
+  // KALDI_ASSERT(initial_state_in_chunk_.size() == 0);
+
+  // step 3.2: connect the states between two chunks
+  if (last_get_lattice_frame_ != 0) {
+    KALDI_ASSERT(final_arc_list_prev_.size());
+    StateId prev_final_state = -1;
+    for (auto &i : final_arc_list_prev_) {
+      MutableArcIterator<CompactLattice> aiter(olat, i.first);
+      aiter.Seek(i.second);
+      // Obtain the appended final arcs in the last chunk
+      auto &arc_appended = aiter.Value();
+      // Find out whether its corresponding Token still exists in the begin
+      // of this chunk. If not, it is pruned by PruneActiveTokens()
+      auto r = initial_arc_map.find(arc_appended.olabel);
+      if (r != initial_arc_map.end()) {
+        ArcIterator<CompactLattice> aiter_unappended(clat, 0); // initial state
+        aiter_unappended.Seek(r->second);
+        const auto &arc_unappended = aiter_unappended.Value();
+        KALDI_ASSERT(arc_unappended.olabel == arc_appended.olabel);
+        StateId state_appended = arc_unappended.nextstate + state_offset;
+        if (prev_final_state == -1)
+          prev_final_state = arc_appended.nextstate;
+        else
+          KALDI_ASSERT(arc_appended.nextstate == prev_final_state);
+        // For the later code in this loop, we try to modify the arc_appended
+        // to connect the last frame state of last chunk to the first frame
+        // state of this chunk. These begin and final states are
+        // corresponding to the same Token, guaranteed by unique state labels.
+        CompactLatticeArc arc_appended_mod(arc_appended);
+        arc_appended_mod.nextstate = state_appended;
+
+        CompactLatticeWeight weight_offset, weight_offset_final;
+        weight_offset.SetWeight(
+            LatticeWeight(0, -state_label_initial_cost_[arc_appended.olabel]));
+        weight_offset_final.SetWeight(
+            LatticeWeight(0, -state_label_final_cost_[arc_appended.olabel]));
+        arc_appended_mod.weight =
+            Times(arc_appended_mod.weight, arc_unappended.weight);
+        arc_appended_mod.weight =
+            Times(arc_appended_mod.weight, olat->Final(prev_final_state));
+        arc_appended_mod.weight = Times(arc_appended_mod.weight, weight_offset);
+        arc_appended_mod.weight =
+            Times(arc_appended_mod.weight, weight_offset_final);
+        arc_appended_mod.olabel = 0;
+        arc_appended_mod.ilabel = 0;
+        aiter.SetValue(arc_appended_mod);
+      } // otherwise, it has been pruned
+      state_label_initial_cost_.erase(arc_appended.olabel);
+      state_label_final_cost_.erase(arc_appended.olabel);
```

Comment:
This is dangerous, because can you guarantee that there will be just a single final-arc per symbol?  I don't think you can.  You might be deleting these costs that will later be needed.  And anyway, why would you delete them one by one if you don't need to?  It's wasteful and unnecessary.

Reply:
Yes, I found that in the determinized lattice, there could be multiple final arcs with the same state label. I need to change the logic here.